tests/unit/s2n_tls13_cert_verify_test.c

```diff
@@ -270,61 +319,135 @@ int run_tests(const struct s2n_tls13_cert_verify_test *test_case, s2n_mode verif
+        uint32_t available_size = s2n_stuffer_data_available(&certificate_out);
+        EXPECT_SUCCESS(s2n_blob_init(&b, s2n_stuffer_raw_read(&certificate_out, available_size),
+                available_size));
         if (verifying_conn->mode == S2N_CLIENT) {
-            EXPECT_SUCCESS(s2n_pkey_free(&verifying_conn->handshake_params.server_public_key));
+            EXPECT_OK(s2n_asn1der_to_public_key_and_type(&verifying_conn->handshake_params.server_public_key,
+                    &pkey_type, &b));
+            EXPECT_SUCCESS(s2n_pkey_match(&verifying_conn->handshake_params.server_public_key,
+                    sending_conn->handshake_params.our_chain_and_key->private_key));
         } else {
-            EXPECT_SUCCESS(s2n_pkey_free(&verifying_conn->handshake_params.client_public_key));
+            EXPECT_OK(s2n_asn1der_to_public_key_and_type(&verifying_conn->handshake_params.client_public_key,
+                    &pkey_type, &b));
+            EXPECT_SUCCESS(s2n_pkey_match(&verifying_conn->handshake_params.client_public_key,
+                    sending_conn->handshake_params.our_chain_and_key->private_key));
         }
 
-        EXPECT_SUCCESS(s2n_stuffer_free(&certificate_in));
-        EXPECT_SUCCESS(s2n_stuffer_free(&certificate_out));
-        EXPECT_SUCCESS(s2n_connection_free(verifying_conn));
+        /* Hash initialization */
+        EXPECT_SUCCESS(s2n_hash_init(&sending_conn->handshake.hashes->sha256, S2N_HASH_SHA256));
+        EXPECT_SUCCESS(s2n_hash_update(&sending_conn->handshake.hashes->sha256, hello, strlen((char *) hello)));
+        EXPECT_SUCCESS(s2n_hash_init(&verifying_conn->handshake.hashes->sha256, S2N_HASH_SHA256));
+        EXPECT_SUCCESS(s2n_hash_update(&verifying_conn->handshake.hashes->sha256, hello, strlen((char *) hello)));
+
+        /* Send cert verify */
+        EXPECT_SUCCESS(s2n_tls13_cert_verify_send(sending_conn));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&sending_conn->handshake.io, &verifying_conn->handshake.io,
+                s2n_stuffer_data_available(&sending_conn->handshake.io)));
+
+        struct s2n_signature_scheme test_scheme = *verifying_conn->handshake_params.server_cert_sig_scheme;
+        verifying_conn->handshake_params.server_cert_sig_scheme = &test_scheme;
+        verifying_conn->handshake_params.client_cert_sig_scheme = &test_scheme;
+
+        /* Use a signature algorithm different from sender */
+        test_scheme.sig_alg = S2N_SIGNATURE_RSA;
+
+        /* Skip the first two bytes (signature scheme) because s2n_tls13_cert_read_and_verify_signature 
+         * expects first two bytes of the message will be the signature size
+         */
+        EXPECT_SUCCESS(s2n_stuffer_skip_read(&verifying_conn->handshake.io, 2));
+
+        EXPECT_FAILURE_WITH_ERRNO(s2n_tls13_cert_read_and_verify_signature(verifying_conn,
+                                          verifying_conn->handshake_params.server_cert_sig_scheme),
```

Comment:
Nit: Given that you have access to the signature scheme parameter with this function, you actually don't need lines 437-438. Instead you can just do:
```suggestion
                                          &test_scheme),
```
Same with your other test where you are calling this specific function.

Reply:
?

Reply:
Addressed this in the latest commit